### PR TITLE
Introduce `Settings::Base`

### PR DESCRIPTION
**Pull request by citizen428**
## What type of PR is this? (check all applicable)

- [X] Refactor

## Description

While working on splitting `SiteConfig` into the various `Settings::*` classes I had to introduce a monkey-patch to allow for class-specific request caches. To avoid breaking our patch we had pinned the gem. It also supports some things we don't really use. For these reasons I had planned to eventually make our own version that we fully own, so this PR introduces the `Settings::Base` class and removes the gem from our Gemfile.

Notable changes:

* I renamed the DSL method from `field` (which never made much sense to me) to `setting`.
* No more complicated monkey-patching, I simplified the whole code by no longer using `ActiveSupport::CurrentAttributes`. Instead, I directly use `RequestStore` which we already had in our `Gemfile`. This has two advantages:
    1. Using a subclass of `ActiveSupport::CurrentAttributes` required some metaprogramming in the `inherited` hook (see the old monkey-patch) and I'd rather do away with that completely.
    2. `RequestStore` doesn't need to be wrapped to provide the same interface as `ActiveSupport::CurrentAttributes` which happens [here](https://github.com/huacnlee/rails-settings-cached/blob/main/lib/rails-settings/request_cache.rb) in the original gem. 
* Cache scopes are gone. We never used them and having to change a key in the source file and redeploy the code to invalidate a cache is impractical. The `clear_cache` class method can be used for clearing caches when necessary, e.g. `Settings::General.clear_cache`.
* Readonly fields have been removed. We never used this feature and it's unlikely that we ever would since we use our own `ApplicationConfig` model for this purpose. This simplifies the code further.
* Removed [instance to class delegation](https://github.com/huacnlee/rails-settings-cached/blob/7695c7f39b6616b7aa6c171425595e749da9955b/lib/rails-settings/base.rb#L131-L136). We never use it.
* Code has been updated to make Rubocop happy.
* I also updated the code to be more consistent with the rest of our codebase, e.g. no `_` prefixes for private methods.

## Related Tickets & Documents

https://github.com/forem/rfcs/pull/126

## QA Instructions, Screenshots, Recordings

* All specs and config related Cypress tests should still be green.
* Manually interact with `/admin/customization/config` and verify that everything works as expected.

### UI accessibility concerns?

n/a

## Added/updated tests?

- [X] Yes

## [Forem core team only] How will this change be communicated?

- [X] I've updated the [Developer Docs](https://docs.forem.com) and/or
      [Admin Guide](https://admin.forem.com/), or
      [Storybook](https://storybook.forem.com/) (for Crayons components)

### CODE REVIEW

File: app/models/settings/base.rb [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+    class ProcetedKeyError < ArgumentError; end
+
+    PROTECTED_KEYS = %w[var value].freeze
+    SEPARATOR_REGEXP = /[\n,;]+/.freeze
+
+    class << self
+      def inherited(subclass)
+        super
+
+        # This defines a custom request cache based on ActiveSupport::CurrentAttributes.
```
Comment (citizen428) [inline app/models/settings/base.rb:22]: This is the old monkey-patch but written in a cleaner and more understandable way. The comment hopefully provides enough context,

File: app/models/settings/base.rb [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+        request_cache = Class.new(ActiveSupport::CurrentAttributes) { attribute :settings }
+        subclass.const_set(:RequestCache, request_cache)
+      end
+
+      def clear_cache
+        self::RequestCache.reset
+        Rails.cache.delete(cache_key)
+      end
+
+      def field(key, default: nil, type: :string, separator: nil, validates: nil)
```
Comment (citizen428) [inline app/models/settings/base.rb:35]: In the original code they signature was `key, **opts` but since this is a public method I opted for making the kwargs explicit.

File: app/models/settings/base.rb [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+          validates: validates,
+        )
+      end
+
+      def get_field(key)
+        @defined_fields.detect { |field| field[:key] == key.to_s } || {}
+      end
+
+      def cache_key
+        @cache_key ||= name.underscore.tr("/", "_")
```
Comment (citizen428) [inline app/models/settings/base.rb:50]: Scope handling (e.g. `v1` was removed here.

File: app/models/settings/base.rb [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+      end
+
+      def keys
+        @defined_fields.pluck(:key)
```
Comment (citizen428) [inline app/models/settings/base.rb:54]: This and similar instances were rewritten to use [`Enumerable#pluck`](https://api.rubyonrails.org/classes/Enumerable.html#method-i-pluck) as suggested by RuboCop.

File: app/models/settings/base.rb [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+      end
+
+      private
+
+      def define_field(key, default: nil, type: :string, separator: nil, validates: nil)
```
Comment (citizen428) [inline app/models/settings/base.rb:59]: The handling of read-only fields was removed from this method.

File: app/models/settings/base.rb [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+          value.to_f
+        when :big_decimal
+          value.to_d
+        else
+          value
+        end
+      end
+
+      def value_of(var_name)
+        unless Database.table_available?(table_name)
```
Comment (citizen428) [inline app/models/settings/base.rb:123]: We already had this code in place so I removed the custom version the gem was using.

**Comment by citizen428** [comment]
There are two major problems at this point in time:

1. Removing `rails-settings-cached` from our `Gemfile` leads to weird `Doorkeeper` errors when trying to start the app:
    ```
    [snip]/config/initializers/doorkeeper.rb:493:in `<main>': uninitialized constant Doorkeeper::AccessGrant
    Did you mean?  Doorkeeper::AccessGrantMixin (NameError)
    ```
    This does not seem related to my branch, I verified that the same happens on current `main`.
2. I have one failing spec in `spec/requests/admin/configs_spec.rb:831`. I can also verify this by performing the actions manually and debugging with `binding.pry`. It's also not tied to my use of `RequestStore`, the same happened when I still used `ActiveSupport::CurrentAttributes`. I don't quite understand why yet but I always get back the old, cached valued of `Settings::General` which doesn't happen with the current code in `main`.

File: app/models/settings/base.rb [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+      end
+
+      def keys
+        @defined_fields.pluck(:key)
+      end
+
+      private
+
+      def cache_key
+        @cache_key ||= name.underscore
```
Comment (citizen428) [inline app/models/settings/base.rb:48]: Scope handling (e.g. `v1`) was removed here.

File: app/models/settings/base.rb [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+        @defined_fields ||= []
+        @defined_fields << {
+          key: key,
+          default: default,
+          type: type || :string
+        }
+
+        # Getter
+        define_singleton_method(key) do
+          result = __send__(:value_of, key)
```
Comment (citizen428) [inline app/models/settings/base.rb:65]: I refactored the code for retrieving the value and generating the default if necessary.

File: app/models/settings/base.rb [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+          value.to_d
+        else
+          value
+        end
+      end
+
+      def value_of(var_name)
+        unless Database.table_available?(table_name)
+          # Fallback to default value if table was not ready (before migrate)
+          Rails.logger.warn("'#{table_name}' does not exist, '#{name}.#{var_name}' will return the default value.")
```
Comment (citizen428) [inline app/models/settings/base.rb:129]: Changed this to use `Rails.logger` instead of `puts`.

File: app/models/settings/base.rb [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+          return
+        end
+
+        all_settings[var_name]
+      end
+
+      def all_settings
+        RequestStore[cache_key] ||= Rails.cache.fetch(cache_key, expires_in: 1.week) do
+          unscoped.select(:var, :value).each_with_object({}) do |record, result|
```
Comment (citizen428) [inline app/models/settings/base.rb:138]: Refactored this whole block.

File: app/models/settings/base.rb [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+# also removed features we don't need like cache scopes and readonly fields.
+#
+# See: https://github.com/huacnlee/rails-settings-cached
+module Settings
+  class Base < ActiveRecord::Base # rubocop:disable Rails/ApplicationRecord
+    self.abstract_class = true
+
+    class SettingNotFound < RuntimeError; end
+
+    class ProcetedKeyError < ArgumentError; end
```
Comment (rhymes) [inline app/models/settings/base.rb:12]: typo:

```suggestion
    class ProtectedKeyError < ArgumentError; end
```

though I see the original gem has this typo too :D https://github.com/huacnlee/rails-settings-cached/blob/7695c7f39b6616b7aa6c171425595e749da9955b/lib/rails-settings/base.rb#L4

Comment (rhymes) [reply app/models/settings/base.rb:12]: should these two classes be inside `Base`? Maybe `Settings::SettingNotFound` instead?

File: app/models/settings/base.rb [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+
+    PROTECTED_KEYS = %w[var value].freeze
+    SEPARATOR_REGEXP = /[\n,;]+/.freeze
+
+    # In rails-settings-cache this is defined in the railtie.
```
Comment (rhymes) [inline app/models/settings/base.rb:17]:
```suggestion
    # In rails-settings-cached this is defined in the railtie.
```

File: app/models/settings/base.rb [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+            result[record.var] = record.value
+          end.with_indifferent_access
+        end
+      end
+    end
+
+    # get the value field, YAML decoded
+    def value
+      YAML.load(self[:value]) if self[:value].present? # rubocop:disable Security/YAMLLoad
+    end
```
Comment (rhymes) [inline app/models/settings/base.rb:148]: what happened to https://github.com/huacnlee/rails-settings-cached/blob/7695c7f39b6616b7aa6c171425595e749da9955b/lib/rails-settings/base.rb#L139 and `YAML.safe_load` ?

**Review by rhymes — COMMENTED** [review]
I'm a little bit wary that by departing from the gem we lose possibile improvements and maybe some testing but I understand why we don't want to live with those patches

**Comment by citizen428** [comment]
> I'm a little bit wary that by departing from the gem we lose possibile improvements

I think I disagree in this specific case. After spending a few hours understanding how the gem works and essentially rewriting parts of it I already removed quite a few features we didn't even use. What remains is short (less than 150 lines of code), simpler (e.g. the cache invalidation callback is now explicit and not hidden in a Railtie), and tailored to what we need, so it's actually much easier to improve because we don't need to wait for others to do it for us.

File: app/models/settings/base.rb (hunk 13 above)
Comment (citizen428) [reply app/models/settings/base.rb:148]: Re `_convert_string_to_typeof_value`: I renamed it to `convert_string_to_value_type` which IMO is a clearer name. That method still uses `YAML.safe_load`:

```ruby
        when :hash
          value = begin
            YAML.safe_load(value).to_h
          rescue StandardError
            {}
          end
          value.deep_stringify_keys!
          ActiveSupport::HashWithIndifferentAccess.new(value)
```

What you're commenting on here is the `value` instance methods. In the version we pinned these were at the bottom of the file, in the current version of the gem they are at the top. I kept them at the bottom because generally Ruby code has instance methods after class methods. As you can see `#value` still uses `YAML.load` in `rails-settings-cached` too:

https://github.com/huacnlee/rails-settings-cached/blob/7695c7f39b6616b7aa6c171425595e749da9955b/lib/rails-settings/base.rb#L18

I experimented with switching this over to `safe_load` but things started breaking. At least the original author got rid of `eval` which was still used in previous versions of the gem.

File: app/models/settings/base.rb (hunk 11 above)
Comment (citizen428) [reply app/models/settings/base.rb:12]: Good call, I moved `ProtectedKeyError` to `Settings`. As for `SettingNotFound`, I removed it completely, it's never used, not even in the gem:

https://github.com/huacnlee/rails-settings-cached/search?q=settingnotfound

File: Gemfile [hunk 14]
```diff
@@ -173,5 +173,6 @@ group :test do
   gem "vcr", "~> 6.0" # Record your test suite's HTTP interactions and replay them during future test runs for fast, deterministic, accurate tests
   gem "webdrivers", "~> 4.6" # Run Selenium tests more easily with install and updates for all supported webdrivers
   gem "webmock", "~> 3.13", require: false # WebMock allows stubbing HTTP requests and setting expectations on HTTP requests
+  gem "with_model", "~> 2.1.5" # Dynamically build a model within an RSpec context
```
Comment (citizen428) [inline Gemfile:176]: I had started to ad-hoc implement something like this on top of SQLite when writing the specs for this PR but thought that surely someone must have already done that. And indeed, there's a very complete gem that allows defining new models ad-hoc in tests: https://github.com/Casecommons/with_model

File: app/models/settings/authentication.rb [hunk 15]
```diff
@@ -1,34 +1,30 @@
...
     self.table_name = :settings_authentications
 
-    # The configuration is cached, change this if you want to force update
-    # the cache, or call Settings::Authentication.clear_cache
-    cache_prefix { "v1" }
-
-    field :allow_email_password_login, type: :boolean, default: true
-    field :allow_email_password_registration, type: :boolean, default: false
-    field :allowed_registration_email_domains, type: :array, default: %w[], validates: {
+    setting :allow_email_password_login, type: :boolean, default: true
```
Comment (citizen428) [inline app/models/settings/authentication.rb:5]: This happened in all the settings files, a simple renaming of the DSL method.

File: spec/models/settings/base_spec.rb [hunk 16]
```diff
@@ -0,0 +1,169 @@
+require "rails_helper"
+
+RSpec.describe Settings::Base, type: :model do
+  with_model :TestSetting, superclass: described_class do
```
Comment (citizen428) [inline spec/models/settings/base_spec.rb:4]: Here we are specifying that we want a new model called `TestSetting` which descends from `Settings::Base`.

File: spec/models/settings/base_spec.rb [hunk 17]
```diff
@@ -0,0 +1,169 @@
...
+    table do |t|
```
Comment (citizen428) [inline spec/models/settings/base_spec.rb:5]: This basically follows the schema DSL.

File: spec/models/settings/base_spec.rb [hunk 18]
```diff
@@ -0,0 +1,169 @@
...
+      t.string :var, null: false
+      t.text :value, null: true
+      t.timestamps
+    end
+
+    model do
```
Comment (citizen428) [inline spec/models/settings/base_spec.rb:11]: Here we put the actual content of our model, e.g. what we would otherwise put in `app/models/test_setting.rb`.

File: config/initializers/doorkeeper.rb [hunk 19]
```diff
@@ -1,3 +1,10 @@
+# Doorkeeper uses ActiveSupport::LazyLoadHooks to lazily load its models.
```
Comment (citizen428) [inline config/initializers/doorkeeper.rb:1]: This one was fun to find.

File: app/models/settings/base.rb (hunk 13 above)
Comment (rhymes) [reply app/models/settings/base.rb]: Got it!

File: config/initializers/doorkeeper.rb (hunk 19 above)
Comment (rhymes) [reply config/initializers/doorkeeper.rb:1]: hahahaah what a weird thing

**Review by rhymes — APPROVED** [review]
LGTM!

File: Gemfile (hunk 14 above)
Comment (joshpuetz) [reply Gemfile:174]: Oh this is really cool!

File: docs/backend/configuration.md [hunk 20]
```diff
@@ -79,3 +62,24 @@ One advantage of using Vault with Forem is that it allows you to update your
 secrets easily through the application rather than having to mess with ENV
 files. If you would like to try out Vault, follow our
 [installation guide for setting it up locally](/installation/vault).
+
+## `Settings::Base`
+
+This is our customized version of the
+[rails-settings-cached](https://github.com/huacnlee/rails-settings-cached) code
+which allows for multiple settings model by defining class-specic request
+caches.
```
Comment (joshpuetz) [inline docs/backend/configuration.md:71]: Suggestion: rephrase this to avoid giving the impression we've forked or are somehow tied to `rails-setting-cached` (which is how I read it). 

```suggestion
This class is adapted from 
[rails-settings-cached](https://github.com/huacnlee/rails-settings-cached).
It allows for multiple settings model by defining class-specic request
caches.
```

**Review by joshpuetz — APPROVED** [review]

File: docs/backend/configuration.md (hunk 20 above)
Comment (citizen428) [reply docs/backend/configuration.md:71]: That's a good point. I actually already have some other changes locally so I won't commit the suggestion as-is but thanks for the feedback @joshpuetz!

File: Gemfile (hunk 14 above)
Comment (citizen428) [reply Gemfile:174]: @joshpuetz It really is. I had something like this:

```ruby
class TestSettings < Settingss::Base
  establish_connection(adapter: "sqlite3", database: "tmp/test_settings.db")

  ...
end
```

That worked but at that point, I realized that I need to be able to define the schema for the backing table and started to wonder how to do this. I'm glad I found the gem because I got about 25% to implementing it before reason took over and I just googled. 😅

File: config/initializers/doorkeeper.rb (hunk 19 above)
Comment (citizen428) [reply config/initializers/doorkeeper.rb:1]: Yes! I thought I was going crazy for a while until I found a random discussion about a similar error in Rake tasks. That led me to `ActiveSupport::LazyLoadHooks` and figuring out that `active_record/base` needs to be required for it to work. I'm jut happy I didn't waste even more time on this.
